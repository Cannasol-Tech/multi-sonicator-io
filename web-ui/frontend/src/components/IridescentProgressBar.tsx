import { useState, useEffect } from 'react'

interface IridescentProgressBarProps {
  progress: number // 0-100
  isRunning: boolean
  isComplete: boolean
  totalItems?: number
  currentItem?: number
  itemLabel?: string
  showStepProgress?: boolean
  stepProgress?: number
  currentStep?: number
  totalSteps?: number
  className?: string
}

export default function IridescentProgressBar({
  progress,
  isRunning,
  isComplete,
  totalItems = 25,
  currentItem,
  itemLabel = 'scenarios',
  showStepProgress = false,
  stepProgress = 0,
  currentStep = 1,
  totalSteps = 10,
  className = ''
}: IridescentProgressBarProps) {
  
  const getProgressBarClass = () => {
    if (isComplete) return 'progress-fill metallic-blue'
    if (isRunning) return 'progress-fill iridescent'
    return 'progress-fill'
  }

  const getStepProgressBarClass = () => {
    if (isComplete) return 'step-progress-fill metallic-blue'
    if (isRunning) return 'step-progress-fill iridescent'
    return 'step-progress-fill'
  }

  const calculatedCurrentItem = currentItem ?? Math.round((progress / 100) * totalItems)

  return (
    <div className={`iridescent-progress-container ${className}`}>
      {/* Main Progress Section */}
      <div className="progress-section">
        <div className="progress-label">
          <span style={{ fontSize: '14px', fontWeight: '500', color: 'var(--text-primary)' }}>
            Test Execution Progress
          </span>
          <span style={{ fontSize: '12px', color: 'var(--text-tertiary)' }}>
            {Math.round(progress)}%
          </span>
        </div>
        
        <div className="progress-bar" style={{
          width: '100%',
          height: '20px',
          background: 'var(--bg-tertiary)',
          borderRadius: '10px',
          overflow: 'hidden',
          marginBottom: '8px'
        }}>
          <div
            className={getProgressBarClass()}
            style={{
              width: `${progress}%`,
              height: '100%',
              background: (!isRunning && !isComplete) ? 'var(--border-color)' : undefined,
              transition: 'width 0.3s ease',
              borderRadius: '10px'
            }}
          />
        </div>
        
        <div style={{ 
          fontSize: '12px', 
          color: 'var(--text-tertiary)', 
          textAlign: 'center',
          marginBottom: showStepProgress ? '15px' : '0'
        }}>
          {calculatedCurrentItem} / {totalItems} {itemLabel} ({Math.round(progress)}%)
        </div>
      </div>

      {/* Step Progress Section */}
      {showStepProgress && isRunning && (
        <div className="step-progress-section">
          <div style={{ 
            fontSize: '11px', 
            color: 'var(--text-tertiary)', 
            marginBottom: '4px' 
          }}>
            Step {currentStep} / {totalSteps}
          </div>
          <div style={{
            width: '100%',
            height: '8px',
            background: 'var(--bg-tertiary)',
            borderRadius: '4px',
            overflow: 'hidden'
          }}>
            <div
              className={getStepProgressBarClass()}
              style={{
                width: `${stepProgress}%`,
                height: '100%',
                background: (!isRunning && !isComplete) ? 'var(--border-color)' : undefined,
                transition: 'width 0.3s ease',
                borderRadius: '4px'
              }}
            />
          </div>
        </div>
      )}

      {/* Status Message */}
      {isRunning && (
        <div style={{
          fontSize: '12px',
          color: 'var(--text-secondary)',
          textAlign: 'center',
          marginTop: '8px',
          fontStyle: 'italic'
        }}>
          ðŸ”„ Test execution in progress...
        </div>
      )}

      {isComplete && (
        <div style={{
          fontSize: '12px',
          color: 'var(--text-secondary)',
          textAlign: 'center',
          marginTop: '8px',
          fontWeight: '600'
        }}>
          âš¡ Test execution complete!
        </div>
      )}
    </div>
  )
}
