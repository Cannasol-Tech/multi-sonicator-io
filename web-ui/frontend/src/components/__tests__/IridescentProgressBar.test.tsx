import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import IridescentProgressBar from '../IridescentProgressBar'

describe('IridescentProgressBar', () => {
  it('should render with basic props', () => {
    render(
      <IridescentProgressBar
        progress={50}
        isRunning={false}
        isComplete={false}
      />
    )
    
    expect(screen.getByText('Test Execution Progress')).toBeInTheDocument()
    expect(screen.getByText('50%')).toBeInTheDocument()
  })

  it('should show running status message when isRunning is true', () => {
    render(
      <IridescentProgressBar
        progress={75}
        isRunning={true}
        isComplete={false}
      />
    )
    
    // Check that the new text is displayed instead of the old "Beautiful iridescent progress in action!"
    expect(screen.getByText('ðŸ”„ Test execution in progress...')).toBeInTheDocument()
    expect(screen.queryByText('ðŸŒˆ Beautiful iridescent progress in action!')).not.toBeInTheDocument()
  })

  it('should show completion status message when isComplete is true', () => {
    render(
      <IridescentProgressBar
        progress={100}
        isRunning={false}
        isComplete={true}
      />
    )
    
    expect(screen.getByText('âš¡ Test execution complete!')).toBeInTheDocument()
  })

  it('should not show status message when not running and not complete', () => {
    render(
      <IridescentProgressBar
        progress={0}
        isRunning={false}
        isComplete={false}
      />
    )
    
    expect(screen.queryByText('ðŸ”„ Test execution in progress...')).not.toBeInTheDocument()
    expect(screen.queryByText('âš¡ Test execution complete!')).not.toBeInTheDocument()
  })

  it('should display progress with custom item labels', () => {
    render(
      <IridescentProgressBar
        progress={60}
        isRunning={false}
        isComplete={false}
        totalItems={10}
        currentItem={6}
        itemLabel="tests"
      />
    )
    
    expect(screen.getByText('6 / 10 tests (60%)')).toBeInTheDocument()
  })

  it('should show step progress when enabled', () => {
    render(
      <IridescentProgressBar
        progress={50}
        isRunning={true}
        isComplete={false}
        showStepProgress={true}
        stepProgress={75}
        currentStep={3}
        totalSteps={4}
      />
    )
    
    expect(screen.getByText('Step 3 / 4')).toBeInTheDocument()
  })

  it('should apply custom className', () => {
    const { container } = render(
      <IridescentProgressBar
        progress={50}
        isRunning={false}
        isComplete={false}
        className="custom-progress"
      />
    )
    
    expect(container.querySelector('.custom-progress')).toBeInTheDocument()
  })

  it('should calculate current item from progress when not provided', () => {
    render(
      <IridescentProgressBar
        progress={40}
        isRunning={false}
        isComplete={false}
        totalItems={25}
      />
    )
    
    // 40% of 25 = 10
    expect(screen.getByText('10 / 25 scenarios (40%)')).toBeInTheDocument()
  })
})
