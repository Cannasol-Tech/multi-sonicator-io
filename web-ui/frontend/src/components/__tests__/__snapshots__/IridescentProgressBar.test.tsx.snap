// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IridescentProgressBar > Visual Regression > matches snapshot for completed state 1`] = `
<div
  class="iridescent-progress-container "
>
  <div
    class="progress-section"
  >
    <div
      class="progress-label"
    >
      <span
        style="font-size: 14px; font-weight: 500; color: var(--text-primary);"
      >
        Test Execution Progress
      </span>
      <span
        style="font-size: 12px; color: var(--text-tertiary);"
      >
        100
        %
      </span>
    </div>
    <div
      class="progress-bar"
      style="width: 100%; height: 20px; background: var(--bg-tertiary); border-radius: 10px; overflow: hidden; margin-bottom: 8px;"
    >
      <div
        class="progress-fill metallic-blue"
        style="width: 100%; height: 100%; transition: width 0.3s ease; border-radius: 10px;"
      />
    </div>
    <div
      style="font-size: 12px; color: var(--text-tertiary); text-align: center; margin-bottom: 0px;"
    >
      10
       / 
      10
       
      tests
       (
      100
      %)
    </div>
  </div>
  <div
    style="font-size: 12px; color: var(--text-secondary); text-align: center; margin-top: 8px; font-weight: 600;"
  >
    âš¡ Test execution complete!
  </div>
</div>
`;

exports[`IridescentProgressBar > Visual Regression > matches snapshot for default idle state 1`] = `
<div
  class="iridescent-progress-container "
>
  <div
    class="progress-section"
  >
    <div
      class="progress-label"
    >
      <span
        style="font-size: 14px; font-weight: 500; color: var(--text-primary);"
      >
        Test Execution Progress
      </span>
      <span
        style="font-size: 12px; color: var(--text-tertiary);"
      >
        42
        %
      </span>
    </div>
    <div
      class="progress-bar"
      style="width: 100%; height: 20px; background: var(--bg-tertiary); border-radius: 10px; overflow: hidden; margin-bottom: 8px;"
    >
      <div
        class="progress-fill"
        style="width: 42%; height: 100%; background: var(--border-color); transition: width 0.3s ease; border-radius: 10px;"
      />
    </div>
    <div
      style="font-size: 12px; color: var(--text-tertiary); text-align: center; margin-bottom: 0px;"
    >
      11
       / 
      25
       
      scenarios
       (
      42
      %)
    </div>
  </div>
</div>
`;

exports[`IridescentProgressBar > Visual Regression > matches snapshot for running state with step progress 1`] = `
<div
  class="iridescent-progress-container "
>
  <div
    class="progress-section"
  >
    <div
      class="progress-label"
    >
      <span
        style="font-size: 14px; font-weight: 500; color: var(--text-primary);"
      >
        Test Execution Progress
      </span>
      <span
        style="font-size: 12px; color: var(--text-tertiary);"
      >
        67
        %
      </span>
    </div>
    <div
      class="progress-bar"
      style="width: 100%; height: 20px; background: var(--bg-tertiary); border-radius: 10px; overflow: hidden; margin-bottom: 8px;"
    >
      <div
        class="progress-fill iridescent"
        style="width: 67%; height: 100%; transition: width 0.3s ease; border-radius: 10px;"
      />
    </div>
    <div
      style="font-size: 12px; color: var(--text-tertiary); text-align: center; margin-bottom: 15px;"
    >
      10
       / 
      15
       
      scenarios
       (
      67
      %)
    </div>
  </div>
  <div
    class="step-progress-section"
  >
    <div
      style="font-size: 11px; color: var(--text-tertiary); margin-bottom: 4px;"
    >
      Step 
      4
       / 
      7
    </div>
    <div
      style="width: 100%; height: 8px; background: var(--bg-tertiary); border-radius: 4px; overflow: hidden;"
    >
      <div
        class="step-progress-fill iridescent"
        style="width: 23%; height: 100%; transition: width 0.3s ease; border-radius: 4px;"
      />
    </div>
  </div>
  <div
    style="font-size: 12px; color: var(--text-secondary); text-align: center; margin-top: 8px; font-style: italic;"
  >
    ðŸ”„ Test execution in progress...
  </div>
</div>
`;
